diff --git a/drivers/net/ovpn/bind.c b/drivers/net/ovpn/bind.c
index 24d2788..1f29e29 100644
--- a/drivers/net/ovpn/bind.c
+++ b/drivers/net/ovpn/bind.c
@@ -48,8 +48,11 @@ struct ovpn_bind *ovpn_bind_from_sockaddr(const struct sockaddr_storage *ss)
  */
 void ovpn_bind_reset(struct ovpn_peer *peer, struct ovpn_bind *new)
 {
+	struct ovpn_bind *old;
+
 	lockdep_assert_held(&peer->lock);
 
-	kfree_rcu(rcu_replace_pointer(peer->bind, new,
-				      lockdep_is_held(&peer->lock)), rcu);
+	old = rcu_replace_pointer(peer->bind, new, lockdep_is_held(&peer->lock));
+	if (old)
+		kfree_rcu(old, rcu);
 }
