diff --git a/drivers/net/ovpn/crypto_aead.c b/drivers/net/ovpn/crypto_aead.c
index 2cca759..ff44ea5 100644
--- a/drivers/net/ovpn/crypto_aead.c
+++ b/drivers/net/ovpn/crypto_aead.c
@@ -36,6 +36,13 @@ static int ovpn_aead_encap_overhead(const struct ovpn_crypto_key_slot *ks)
 		crypto_aead_authsize(ks->encrypt);	/* Auth Tag */
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 3, 0)
+static inline void ovpn_encrypt_post_compl(struct crypto_async_request *req, int ret)
+{
+	ovpn_encrypt_post(req->data, ret);
+}
+#endif
+
 int ovpn_aead_encrypt(struct ovpn_peer *peer, struct ovpn_crypto_key_slot *ks,
 		      struct sk_buff *skb)
 {
@@ -138,7 +145,11 @@ int ovpn_aead_encrypt(struct ovpn_peer *peer, struct ovpn_crypto_key_slot *ks,
 
 	/* setup async crypto operation */
 	aead_request_set_tfm(req, ks->encrypt);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 3, 0)
+	aead_request_set_callback(req, 0, ovpn_encrypt_post_compl, skb);
+#else
 	aead_request_set_callback(req, 0, ovpn_encrypt_post, skb);
+#endif
 	aead_request_set_crypt(req, sg, sg,
 			       skb->len - ovpn_aead_encap_overhead(ks), iv);
 	aead_request_set_ad(req, OVPN_AAD_SIZE);
@@ -147,6 +158,13 @@ int ovpn_aead_encrypt(struct ovpn_peer *peer, struct ovpn_crypto_key_slot *ks,
 	return crypto_aead_encrypt(req);
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 3, 0)
+static inline void ovpn_decrypt_post_compl(struct crypto_async_request *req, int ret)
+{
+	ovpn_decrypt_post(req->data, ret);
+}
+#endif
+
 int ovpn_aead_decrypt(struct ovpn_peer *peer, struct ovpn_crypto_key_slot *ks,
 		      struct sk_buff *skb)
 {
@@ -233,7 +251,11 @@ int ovpn_aead_decrypt(struct ovpn_peer *peer, struct ovpn_crypto_key_slot *ks,
 
 	/* setup async crypto operation */
 	aead_request_set_tfm(req, ks->decrypt);
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 3, 0)
+	aead_request_set_callback(req, 0, ovpn_decrypt_post_compl, skb);
+#else
 	aead_request_set_callback(req, 0, ovpn_decrypt_post, skb);
+#endif
 	aead_request_set_crypt(req, sg, sg, payload_len + tag_size, iv);
 
 	aead_request_set_ad(req, OVPN_AAD_SIZE);
