diff --git a/drivers/net/ovpn/peer.c b/drivers/net/ovpn/peer.c
index 26b55d8..8ff61fe 100644
--- a/drivers/net/ovpn/peer.c
+++ b/drivers/net/ovpn/peer.c
@@ -416,7 +416,7 @@ static __be32 ovpn_nexthop_from_skb4(struct sk_buff *skb)
 {
 	const struct rtable *rt = skb_rtable(skb);
 
-	if (rt && rt->rt_uses_gateway)
+	if (rt && ovpn_rt4_uses_gateway(rt))
 		return rt->rt_gw4;
 
 	return ip_hdr(skb)->daddr;
@@ -797,7 +797,7 @@ static __be32 ovpn_nexthop_from_rt4(struct ovpn_priv *ovpn, __be32 dest)
 		return dest;
 	}
 
-	if (!rt->rt_uses_gateway)
+	if (!ovpn_rt4_uses_gateway(rt))
 		goto out;
 
 	dest = rt->rt_gw4;
@@ -827,8 +827,7 @@ static struct in6_addr ovpn_nexthop_from_rt6(struct ovpn_priv *ovpn,
 		.daddr = dest,
 	};
 
-	entry = ipv6_stub->ipv6_dst_lookup_flow(dev_net(ovpn->dev), NULL, &fl,
-						NULL);
+	entry = ovpn_ipv6_dst_lookup_flow(dev_net(ovpn->dev), NULL, &fl, NULL);
 	if (IS_ERR(entry)) {
 		net_dbg_ratelimited("%s: no route to host %pI6c\n",
 				    netdev_name(ovpn->dev), &dest);
