diff --git a/drivers/net/ovpn/tcp.c b/drivers/net/ovpn/tcp.c
index 0d7f303..59fe382 100644
--- a/drivers/net/ovpn/tcp.c
+++ b/drivers/net/ovpn/tcp.c
@@ -10,6 +10,7 @@
 #include <net/hotdata.h>
 #include <net/inet_common.h>
 #include <net/ipv6.h>
+#include <net/sock.h>
 #include <net/tcp.h>
 #include <net/transp_v6.h>
 #include <net/route.h>
@@ -30,8 +31,14 @@
 
 static struct proto ovpn_tcp_prot __ro_after_init;
 static struct proto_ops ovpn_tcp_ops __ro_after_init;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 16, 0)
+static struct proto ovpn_tcp6_prot;
+static struct proto_ops ovpn_tcp6_ops;
+static DEFINE_MUTEX(tcp6_prot_mutex);
+#else
 static struct proto ovpn_tcp6_prot __ro_after_init;
 static struct proto_ops ovpn_tcp6_ops __ro_after_init;
+#endif
 
 static int ovpn_tcp_parse(struct strparser *strp, struct sk_buff *skb)
 {
@@ -138,14 +145,23 @@ static void ovpn_tcp_rcv(struct strparser *strp, struct sk_buff *skb)
 	kfree_skb(skb);
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 19, 0)
+static int ovpn_tcp_recvmsg(struct sock *sk, struct msghdr *msg, size_t len,
+			    int noblock, int flags, int *addr_len)
+#else
 static int ovpn_tcp_recvmsg(struct sock *sk, struct msghdr *msg, size_t len,
 			    int flags, int *addr_len)
+#endif
 {
 	int err = 0, off, copied = 0, ret;
 	struct ovpn_socket *sock;
 	struct ovpn_peer *peer;
 	struct sk_buff *skb;
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 19, 0)
+	flags |= noblock ? MSG_DONTWAIT : 0;
+#endif /* LINUX_VERSION_CODE < KERNEL_VERSION(5, 19, 0) */
+
 	rcu_read_lock();
 	sock = rcu_dereference_sk_user_data(sk);
 	if (unlikely(!sock || !sock->peer || !ovpn_peer_hold(sock->peer))) {
@@ -155,7 +171,8 @@ static int ovpn_tcp_recvmsg(struct sock *sk, struct msghdr *msg, size_t len,
 	peer = sock->peer;
 	rcu_read_unlock();
 
-	skb = __skb_recv_datagram(sk, &peer->tcp.user_queue, flags, &off, &err);
+	skb = ovpn_skb_recv_datagram(sk, &peer->tcp.user_queue, flags, &off,
+				     &err);
 	if (!skb) {
 		if (err == -EAGAIN && sk->sk_shutdown & RCV_SHUTDOWN) {
 			ret = 0;
@@ -321,7 +338,7 @@ void ovpn_tcp_send_skb(struct ovpn_peer *peer, struct sock *sk,
 	spin_lock_nested(&sk->sk_lock.slock, OVPN_TCP_DEPTH_NESTING);
 	if (sock_owned_by_user(sk)) {
 		if (skb_queue_len(&peer->tcp.out_queue) >=
-		    READ_ONCE(net_hotdata.max_backlog)) {
+		    ovpn_tcp_max_backlog()) {
 			dev_dstats_tx_dropped(peer->ovpn->dev);
 			kfree_skb(skb);
 			goto unlock;
@@ -524,13 +541,21 @@ int ovpn_tcp_socket_attach(struct ovpn_socket *ovpn_sock,
 		ovpn_sock->sk->sk_prot = &ovpn_tcp_prot;
 		ovpn_sock->sk->sk_socket->ops = &ovpn_tcp_ops;
 	} else {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 16, 0)
+		mutex_lock(&tcp6_prot_mutex);
+		if (!ovpn_tcp6_prot.recvmsg)
+			ovpn_tcp_build_protos(&ovpn_tcp6_prot, &ovpn_tcp6_ops,
+					      ovpn_sock->sk->sk_prot,
+					      ovpn_sock->sk->sk_socket->ops);
+		mutex_unlock(&tcp6_prot_mutex);
+#endif
 		ovpn_sock->sk->sk_prot = &ovpn_tcp6_prot;
 		ovpn_sock->sk->sk_socket->ops = &ovpn_tcp6_ops;
 	}
 
 	/* avoid using task_frag */
 	ovpn_sock->sk->sk_allocation = GFP_ATOMIC;
-	ovpn_sock->sk->sk_use_task_frag = false;
+	ovpn_sock_disable_task_frag(ovpn_sock->sk);
 
 	/* enqueue the RX worker */
 	strp_check_rcv(&peer->tcp.strp);
@@ -612,8 +637,10 @@ void __init ovpn_tcp_init(void)
 	ovpn_tcp_build_protos(&ovpn_tcp_prot, &ovpn_tcp_ops, &tcp_prot,
 			      &inet_stream_ops);
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 16, 0)
 #if IS_ENABLED(CONFIG_IPV6)
 	ovpn_tcp_build_protos(&ovpn_tcp6_prot, &ovpn_tcp6_ops, &tcpv6_prot,
 			      &inet6_stream_ops);
 #endif
+#endif /* LINUX_VERSION_CODE >= KERNEL_VERSION(6, 16, 0) */
 }
