diff --git a/tools/net/ynl/pyynl/cli.py b/tools/net/ynl/pyynl/cli.py
index 602ff68c23e0..fc9a56170091 100755
--- a/tools/net/ynl/pyynl/cli.py
+++ b/tools/net/ynl/pyynl/cli.py
@@ -261,7 +261,10 @@ def main():
     dbg_group.add_argument('--dbg-small-recv', default=0, const=4000,
                            action='store', nargs='?', type=int, metavar='INT',
                            help="Length of buffers used for recv()")
-    dbg_group.add_argument('--process-unknown', action=argparse.BooleanOptionalAction)
+    if hasattr(argparse, 'BooleanOptionalAction'):
+        dbg_group.add_argument('--process-unknown', action=argparse.BooleanOptionalAction)
+    else:
+        dbg_group.add_argument('--process-unknown', action='store_true')
 
     args = parser.parse_args()
 
@@ -274,7 +277,7 @@ def main():
 
     if args.list_families:
         for filename in sorted(os.listdir(spec_dir())):
             if filename.endswith('.yaml'):
-                print(filename.removesuffix('.yaml'))
+                print(filename[:-5])
         return
 
     if args.no_schema:
diff --git a/tools/net/ynl/pyynl/lib/nlspec.py b/tools/net/ynl/pyynl/lib/nlspec.py
index e92a103e3c57..d374fbb7a561 100644
--- a/tools/net/ynl/pyynl/lib/nlspec.py
+++ b/tools/net/ynl/pyynl/lib/nlspec.py
@@ -227,7 +227,8 @@ class SpecAttrSet(SpecElement):
         else:
             real_set = family.attr_sets[self.subset_of]
             for elem in self.yaml['attributes']:
                 real_attr = real_set[elem['name']]
-                combined_elem = real_attr.yaml | elem
+                combined_elem = real_attr.yaml.copy()
+                combined_elem.update(elem)
                 attr = self.new_attr(combined_elem, real_attr.value)
 
                 self.attrs[attr.name] = attr
